R/Spectra.R

```diff
@@ -605,30 +684,46 @@ setMethod("Spectra", "MsBackend", function(object, processingQueue = list(),
         backend = object)
 })
 
+#' @rdname Spectra
+setMethod("Spectra", "character", function(object, processingQueue = list(),
+                                           metadata = list(),
+                                           source = MsBackendMzR(),
+                                           backend = MsBackendDataFrame(),
+                                           ..., BPPARAM = bpparam()) {
+    be <- backendInitialize(source, object, ..., BPPARAM = BPPARAM)
+    sp <- new("Spectra", metadata = metadata, processingQueue = processingQueue,
+              backend = be)
+    if (class(source) != class(backend))
+        setBackend(sp, backend, ..., BPPARAM = BPPARAM)
+    else sp
+})
+
 #' @rdname Spectra
 #'
 #' @exportMethod setBackend
 setMethod("setBackend", c("Spectra", "MsBackend"),
-          function(object, backend, f = fromFile(object), ...,
+          function(object, backend, f = dataStorage(object), ...,
                    BPPARAM = bpparam()) {
               backend_class <- class(object@backend)
-              f <- factor(f, levels = unique(f))
+              if (isReadOnly(backend))
+                  stop(backend_class, " is read-only. Changing backend to a ",
+                       "read-only backend is not supported.")
+              f <- force(factor(f, levels = unique(f)))
               if (length(f) != length(object))
                   stop("length of 'f' has to match the length of 'object'")
+              data_storage <- object@backend$dataStorage
               bknds <- bplapply(split(object@backend, f = f), function(z, ...) {
-                  if (isReadOnly(backend) && any(z@modCount))
-                      stop(class(backend), " backends are read-only but the ",
-                           "original data appears to be modified")
-                  res <- backendInitialize(backend, files = z@files,
-                                           spectraData = spectraData(z),
-                                           ...)
-                  res@modCount <- z@modCount
-                  res
+                  backendInitialize(backend,
+                                    spectraData = spectraData(z),
+                                    ...)
               }, ..., BPPARAM = BPPARAM)
               bknds <- backendMerge(bknds)
+              ## That below ensures the backend is returned in its original
+              ## order - unsplit does unfortunately not work.
               if (is.unsorted(f))
                   bknds <- bknds[order(unlist(split(seq_along(bknds), f),
```

Comment:
I don't get this. Didn't this completely destroy the order?
```r
x <- 1:10
f <- rep(c(3, 2, 1, 4, 5), 2)
x
# [1]  1  2  3  4  5  6  7  8  9 10
unsplit(split(x, f), f)
# [1]  1  2  3  4  5  6  7  8  9 10
is.unsorted(f)
# [1] TRUE

order(unlist(split(seq_along(x), f)))
#  [1]  5  3  1  7  9  6  4  2  8 10
x[order(unlist(split(seq_along(x), f)))]
#  [1]  5  3  1  7  9  6  4  2  8 10
```

Reply:
I have to have a look into this again.

Reply:
You forgot the `split(x, f)` in your example above.

```
unlist(split(x, f), use.names = FALSE)[order(unlist(split(seq_along(x), f)))]
[1] 1  2  3  4  5  6  7  8  9 10
```

The reason for this clumsy code is that `unsplit` does not work on `DataFrame`. Happy to change if you find a better solution.

Reply:
Sorry, didn't know that this is different to `base::data.frame`. Keep it that way.